### ISSUE TITLE
Introduce #dsoHandle and args in addition #file, #function and #line

### ISSUE THREAD

**Issue by shpakovski** [apple/swift-log#72]
Hello,

This is more like a pitch, not a an issue at all, and it’s probably too late. However, I’d like to have a custom handler for `os_log`, something like this works fine in Debug code, even though it’s using private API:

```swift
import os
import _SwiftOSOverlayShims

...

func log(
  level: Logger.Level, message: Logger.Message, metadata: Logger.Metadata?,
  file: String, function: String, line: UInt, dso: UnsafeRawPointer, args: CVarArg...)
{
  let ra = _swift_os_log_return_address()
  message.withUTF8Buffer { (buf: UnsafeBufferPointer<UInt8>) in
    buf.baseAddress?.withMemoryRebound(to: CChar.self, capacity: buf.count) { str in
      withVaList(args) { valist in
        _swift_os_log(dso, ra, self.log_from_label(), self.type_from_level(level), message, valist)
      }
    }
  }
}
```

However, the current `log` signature misses `#dsoHandle` and `args`, using `#file`, `#function` and `#line` instead. What do you think about adding two more arguments to the `log` function? Or maybe you have some alternative suggestion?

Thanks in advance!

**Comment by weissi** [comment]
@shpakovski Thanks for the issue. The reason I didn't add `#dsoHandle` is that with SwiftPM it almost always be the same for everything because SwiftPM by default does 'automatic linkage' which means it links it all together in one binary. Ie. it doesn't produce any `.dylib`/`.so` or `.a` files which is a good thing. Now I do realise that especially non-server people use `.frameworks` and then the `#dsoHandle` isn't always the same.

So if you use SwiftPM, you can just pass `#dso` of your current file to `_swift_os_log` :)

**Comment by shpakovski** [comment]
I need this for a Mac app which is full of frameworks 🙈 In such case, `#dsoHandle` also “brings” a location in source code, but I don’t know how exactly that magic works.

I’m not sure about this pitch anyway. Because of `os_log` args and `public` / `private` modifiers these systems cannot be married probably. So I was just curious about the future compatibility.

It’s great that you replied with more details. And thanks a lot for consideration!

**Comment by DeFrenZ** [comment]
I agree that interfacing with `os_log` through private APIs should be discouraged, and as such not supported by `swift-log` out of the box, but it should be possible to interface with it by writing the handler ourselves.
There are a few limits in the current interface that makes that impossible:
- there is no way to pass the `#dsoHandle`, as mentioned upthread
- the only data passed is `Message` (which is basically a `String`), and optional `Metadata` (keyed strings, with possible nesting). This leads to the impossibility of knowing at runtime wether — and if so which parts — of the message were static and which were dynamic (e.g. `StaticString`). Without this knowledge, functionality like the automatic redaction of dynamic elements cannot be used, and even if reimplemented you'd need a different `Logger` to be able to pass it through.

I did in a private project re-use a modified version of `swift-log` with workarounds such as passing a format and arguments instead of a `Message`, and having multiple versions of `log` as requirements instead to "solve" the variadic propagation issue, but I'd love to be able to use `swift-log` as-is without losing functionality like you would with [current solutions](https://github.com/chrisaljoudi/swift-log-oslog/blob/4bfc8f5d924a640f1367469f9d18d31283cbbfe2/Sources/LoggingOSLog/LoggingOSLog.swift#L33).

Though as I understand there's no desire to break the API for this added functionality, with the chance that once a "Swifty" version of `os_log` comes forward might be incompatible 😞
